Fix tools/windows_buildenv.bat and use 7z/PowerShell for extraction

Holzhaus opened the pull request:
Patches from Zulip: https://mixxx.zulipchat.com/#narrow/stream/247620-development-help/topic/compiling.20on.20windows.20help/near/242252112

**EDIT** This makes it possible to build Mixxx from the command line

TODO:
- [ ] Fix missing `pthreadVC3.dll` issue (not in this PR)

JosepMaJAZ committed:
tools/windows_buildenv.bar: Fix setting of buildenv variables

JosepMaJAZ committed:
tools/windows_buildenv.bat: Use 7z or powershell for unzipping

Holzhaus commented:
Untested. @JosepMaJAZ Is PowerShell present on all windows systems by default? Or at least when VS is installed? Otherwise extraction might fail.

JoergAtGithub commented:
I can't execute this:
![grafik](https://user-images.githubusercontent.com/64457745/121601916-dbdac080-ca46-11eb-8e87-4a05a3643da1.png)
The keyword EQ (equal) is not recognized. Also my editor (Notepad++) does no syntax highlighting for this keyword.

Holzhaus commented on tools/windows_buildenv.bat:
```diff
@@ -62,13 +65,17 @@ EXIT /B 0
         ECHO ### Unpacking !BUILDENV_PATH!.zip ###
-        CALL :UNZIP "!BUILDENV_PATH!.zip" "!BUILDENV_BASEPATH!"
+        CALL :DETECT_7ZIP
+        IF "%RETVAL%" EQ "" (
```
```suggestion
        IF "%RETVAL%" == "" (
```
@JoergAtGithub Would this work?

JoergAtGithub replied:
This seems to work 
![grafik](https://user-images.githubusercontent.com/64457745/121603674-6c1a0500-ca49-11eb-81a5-ba3611b4f0bb.png)

Holzhaus replied:
```suggestion
        IF NOT DEFINED RETVAL (
```

JoergAtGithub replied:
The batch file just suceeded with the Powershell version (with optional installation of Powershell version 5.1 under Win7).

Holzhaus replied:
So powershell is not installed on Win7 by default? But I don't think that's a problem, Win7 reached EOL 1.5 years ago.

JoergAtGithub replied:
It comes with an older version of Powershell. I tested only with the newer 5.1 powershell.

JoergAtGithub commented:
The last commit seems to contain unintended character changes

Holzhaus commented on tools/windows_buildenv.bat:
```diff
@@ -65,7 +65,12 @@ EXIT /B 0
-        CALL :UNZIP "!BUILDENV_PATH!.zip" "!BUILDENV_BASEPATH!"
+        CALL :DETECT_7ZIP
+        IF NOT DEFINED RETVAL (
+            CALL :UNZIP "!UNZIP_POWERSHELL!.zip" "!BUILDENV_BASEPATH!"
```
@JosepMaJAZ This looks weird, are you sure this is correct? it looks like it should say:

```suggestion
            CALL :UNZIP_POWERSHELL "!BUILDENV_PATH!.zip" "!BUILDENV_BASEPATH!"
````

Holzhaus commented:
> The last commit seems to contain unintended character changes

fixed.

Holzhaus commented:
I checked with a clean VM. Unfortunately, this doesn't work. I see this on the command line:
```
'7z.exe' is not recognized as an internal or external command, operable program or batch file.
The system cannot find the path specified.
The system cannot find the path specified.
'""""' is not recognized as an internal or external command, operable program or batch file.  
```

Holzhaus commented on tools/windows_buildenv.bat:
```diff
@@ -62,13 +65,17 @@ EXIT /B 0
         ECHO ### Unpacking !BUILDENV_PATH!.zip ###
-        CALL :UNZIP "!BUILDENV_PATH!.zip" "!BUILDENV_BASEPATH!"
+        CALL :DETECT_7ZIP
+        IF NOT DEFINED RETVAL (
```
```suggestion
        IF [%RETVAL%]==[] (
```

Holzhaus committed:
README: Fix name of x64 nativate tools prompt

Holzhaus committed:
tools/windows_buildenv: Improve command line output

Holzhaus committed:
tools/windows_buildenv.bat: Fix CMAKE_PREFIX_PATH

Holzhaus committed:
CMake: Read env vars from buildenv script

This fixes building Mixxx after running the windows buildenv script with
just `cmake -G Ninja ..` instead of requiring to pass various cmake
args.

Holzhaus commented:
Ok, this finally works.

On a fresh VM, I can follow the instructions from the README by the letter and build Mixxx. The only prerequisite is installing VS 2019 Community Edition and ticking the "Desktop development with C++" and "Git for Windows" components in the installer.

Holzhaus committed:
tools/windows_buildenv.bat: Set Ninja generator from buildenv script

This makes the build instructions on all supported OSes
(Linux/macOS/Windows) identical. No special cases anymore.

Holzhaus committed:
CI: Remove generator from CMake command line

This is not necessary anymore. If it fails, our buildenv scripts are
insufficient, so this also helps testing these scripts.

Holzhaus committed:
CI: Remove unused QT_PATH env var from command line

Holzhaus committed:
README: Clarify build instructions

Holzhaus committed:
gitignore: Also ignore backup versions of CMakeSettings.json

JoergAtGithub commented:
The buildenv script succeeded, but cmake .. fails. It doesn't find qt anymore:
![grafik](https://user-images.githubusercontent.com/64457745/121715316-01b3a400-cadf-11eb-8dcd-41d6f5a66cf2.png)

Holzhaus commented:
Did you try with a clean build dir and fresh native tools prompt? It works for me.

JoergAtGithub commented:
CMake hangs now for ~3 minutes between these two lines:
-- Support for sccache: OFF
-- Could NOT find Python (missing: Python_EXECUTABLE Interpreter)
but continious without error. 
This 3 minute wait happen also, if I execute CMake from within VisualStudio, but there is one printout more, before the wait: 1> [CMake] -- Support for sccache: OFF
1> [CMake] -- DEBUG_ASSERT statements have been enabled because DEBUG_ASSERTIONS_FATAL is ON.
Before this PR, I never expirienced a noticeable wait during CMake.

JoergAtGithub commented:
> 
> 
> Did you try with a clean build dir and fresh native tools prompt? It works for me.

Yes, I deleted build and buildenv completely!

JoergAtGithub commented:
I tried everything again, same result, but I noticed, that the failing CMake from command line sets x86 optimization
![grafik](https://user-images.githubusercontent.com/64457745/121721210-4988fa00-cae4-11eb-9cf3-1d5c2b0e610d.png)
while the successful VisualStudio build sets x64 optimization
![grafik](https://user-images.githubusercontent.com/64457745/121721336-71785d80-cae4-11eb-9474-b7088c063861.png)

Holzhaus committed:
CMake: Fix Qt5Keychain detection

JoergAtGithub commented:
It seems I used the wrong command line prompt - will try again

JoergAtGithub commented:
No messages about Qt5Keychain anymore, neither success nor error.
cmake .. succeeded now. Still with the 3 minute hang.

Holzhaus commented:
> No messages about Qt5Keychain anymore, neither success nor error.
> cmake .. succeeded now. Still with the 3 minute hang. 

But apart from that it works? I have no idea where the hang comes from.

Holzhaus committed:
CMake: Fix setting of X_VCPKG_APPLOCAL_DEPS_INSTALL

JoergAtGithub commented:
> 
> 
> > No messages about Qt5Keychain anymore, neither success nor error.
> > cmake .. succeeded now. Still with the 3 minute hang.
> 
> But apart from that it works?

I had not the time for a full build yet, but it started to build.

Holzhaus commented:
Finally all green now. Please test.

JoergAtGithub commented:
Will do tomorrow

JoergAtGithub commented:
I did a first test:
-buildenv script works with 7zip
-Command line configure succeeds (still with the 3 minutes hang)
-Command line build also succeeds
It creates no subdirectory for the build type (x64__portable) in build directory. The following runtime dependencies is not found and mixxx.exe can't be executed:
![grafik](https://user-images.githubusercontent.com/64457745/121769692-61ed2900-cb65-11eb-96a9-dcdad507d262.png)
The DLL exists in the buildenv, but is not copied to the build directory.

I will test later today:
If the VisualStudio build is still working with this PR.
And the case without 7zip.

JoergAtGithub commented:
Now I repeated the steps with VisualStudio. Everythink succeeded and the ./build/x64_portable directory still contains all DLLs to run mixxx.exe
But the CMake configure step also hangs for about 3 minutes in this environment.

JoergAtGithub commented:
@JosepMaJAZ I deinstalled 7zip and run the Powershell Unzip. The process finished, but reported two files as missing:
![grafik](https://user-images.githubusercontent.com/64457745/121774255-e3eb4b00-cb81-11eb-8671-8bd5529cafa9.png)
a directory compare shows that these files were existing in the buildenv extracted by 7zip: 
![grafik](https://user-images.githubusercontent.com/64457745/121774399-97543f80-cb82-11eb-8334-2e7dd07422a0.png)

JoergAtGithub commented:
We have two working unzip methods. The old one which is slow and the 7zip version which is the fastest. If we cant fix the powershell version, I recommend to use the proven old unzip method as fallback for 7zip.

Holzhaus commented:
I'm a bit at a loss. There are two issues left:

1. Long wait time between detecting sccache and python
2. `pthreadsVC3.dll` is not copied whne running cmake from the command line.

Regarding 1: are you sure that the wait time only occurs with this PR? I don't see a difference in the generated CMakeSettings.json, so I honestly dont know why this PR should change anything for running cmake from VS2019. Can you diff the generated cmakesettings.json files?

Holzhaus commented:
> We have two working unzip methods. The old one which is slow and the 7zip version which is the fastest. If we cant fix the powershell version, I recommend to use the proven old unzip method as fallback for 7zip.

Ooof, the powershell version works for me btw. But I guess I'll just revert it completely, I don't want to debug that now. Feel free to post a follow up PR that re-adds the 7zip extraction.

JoergAtGithub commented:
> 
> Regarding 1: are you sure that the wait time only occurs with this PR? I don't see a difference in the generated CMakeSettings.json, so I honestly dont know why this PR should change anything for running cmake from VS2019. Can you diff the generated cmakesettings.json files?

I compared CMakeSettings.json with an older one, they are identical. But the hang is there with VisualStudio as well as with Command Line CMake.
I have to cross check with plain 2.3, maybe this regression was introduced earlier and is not related to this PR.

JoergAtGithub commented:
> 
> 
> > We have two working unzip methods. The old one which is slow and the 7zip version which is the fastest. If we cant fix the powershell version, I recommend to use the proven old unzip method as fallback for 7zip.
> 
> Ooof, the powershell version works for me btw. But I guess I'll just revert it completely, I don't want to debug that now. Feel free to post a follow up PR that re-adds the 7zip extraction.

I can do this!

JoergAtGithub commented:
> 
> Ooof, the powershell version works for me btw. But I guess I'll just revert it completely, I don't want to debug that now. Feel free to post a follow up PR that re-adds the 7zip extraction.

I fear that was too early. I just executed the script from plain 2.3 and ran into this:
![grafik](https://user-images.githubusercontent.com/64457745/121776718-bc4eaf80-cb8e-11eb-9477-05b8b1cdec82.png)
I never saw this before, but now it happens with powershell AND the old unzip.

Holzhaus commented:
Haha, the old windows path length issue. I suggest to put the mixxx code to C:\mixxx, not to C:\some\very\long\path\mixxx.

JoergAtGithub commented:
But I wonder which path? It can't be the target path of the buildenv itself, because it was exactly the same, as used for the 7zip test which succeeded before.

Holzhaus commented:
No idea. IIRC there is some way to increase the max path on windows or something.

JoergAtGithub commented:
I test 2.3 now, to see if the 3 minute hang occurs there too. Than I will go back to this PR.

JoergAtGithub commented:
I just tested plain 2.3. The 3 minute wait during CMake configure also happens there. Therefore it's not related to this PR!

Holzhaus commented:
Does the pthreadVC3.dll issue also happen on 2.3? Because I honestly have no idea what causes it.

JoergAtGithub commented:
The pthreadVC3.dll occurs only with the build from command line. Which was not possible before this PR.

Holzhaus commented:
If this doesn't break anything, this should be ready to merge now. That weird pthreadV3.dll issue should be dealt with separately. It's apparently related to our custom libshout-idjc hack, because with `-DBROADCAST=OFF` it works fine.

JoergAtGithub commented:
I think I found part of the reason for the pthread issue. I will debug it this evening 
I confirm, that this doesn't break anything!

Holzhaus commented:
One way to check would be to compare the output of `cmake -LA` from command line vs. `cmake -LA` on the build dir generated by VS.

JoergAtGithub commented:
I found that there are two issues:
1.) Builds using the unoptimized debug DLLs from buildenv have always the pthread issue (also with VisualStudio build)
2.) In case of command line build, a mixture of an optimized (portable) mixxx.exe and unoptimized debug DLLs from the buildenv is used.

For 2.) I've a fix, where I can create a follow-up PR

Holzhaus commented:
In that case let's merge this now, and take care of the other issues in a follow up PR. These were probably not caused by this PR.

JoergAtGithub committed:
CMake: Set default build type if not set explicitly

JoergAtGithub commented:
This code is not correct, it will not fix the build, as it set never RelWithDebInfo as needed!

And the section if(EXISTS "${CMAKE_SOURCE_DIR}/.git") will result in the same as we had before - local builds will always fail.

Why using this untested code, instead of the snipped I tested for hours in all relevant combinations?

Holzhaus commented:
Wait, I thought the issue was that CMAKE_BUILD_TYPE was not set explicitly. Are you saying that debug builds on windows are entirely broken?

JoergAtGithub commented:
Yes, this is the problem with pthread DLL. This affects all "Debug" builds, but not "RelWithDebInfo". Because CMake defaults to "Debug", all command line builds were failing.

Holzhaus commented:
> Yes, this is the problem with pthread DLL. This affects all "Debug" builds, but not "RelWithDebInfo". Because CMake defaults to "Debug", all command line builds were failing.

Hmm that sucks. We should probably add debug builds to our CI builds to avoid stuff like that in the future. But I have an idea what could be the cause, let me check something.

Holzhaus commented:
OK, fixing the windows debug builds will take some more work. I'll move that to another PR.

Holzhaus committed:
CMake: Default to RelWithDebInfo builds on Windows

Holzhaus commented:
Ok, let's default to `RelWithDebInfo` on windows as a workaround. Ready to merge IMHO.

daschuer commented on CMakeLists.txt:
```diff
@@ -25,6 +25,39 @@ if(POLICY CMP0071)
+  message(STATUS "Setting CMAKE_BUILD_TYPE to '${default_build_type}' as none was specified.")
+  set(CMAKE_BUILD_TYPE "${default_build_type}" CACHE STRING "Choose the type of build." FORCE)
+  # Set the possible values of build type for cmake-gui
+  set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release" "MinSizeRel" "RelWithDebInfo")
```
MinSizeRel is not maintained, remove?

Holzhaus replied:
It's one of the standard values supported by cmake. No idea if it works or not.

daschuer replied:
It does not work, because it implies to use -Os, but we don't use the build type to set the optimization but the optimize parameter. "MinSizeRel" is not considered in this code. 

I just noticed that even "Release" is not proper handled everywhere. 
The later should be fixed because it is an often used build type. 
I can take care or will you? It is just a matter of check both types.

Holzhaus replied:
Please do, you already looked into it. I suggest to merge this PR first, then follow up with more cleanup in another PR. At least this fixes the windows cli build.

ywwg commented:
I had to delete my existing build/ and buildenv/ directories, and then it worked!

JoergAtGithub commented on CMakeLists.txt:
```diff
@@ -25,6 +25,39 @@ if(POLICY CMP0071)
+# Set a default build type if none was specified
+# See https://blog.kitware.com/cmake-and-the-default-build-type/ for details.
+set(default_build_type "Release")
+if(EXISTS "${CMAKE_SOURCE_DIR}/.git")
```
This seems not to match:
1.) Mixxx for Windows are not build as Release until now (Even the official builds on the download page are all "RelWithDebInfo"). Why should the local build use "Release" if the published version doesn't uses it.
2.) IMHO the dependency on .git is fishy. Why this automatic for one single parameter? Others, like Optimize need to be changed manually as well.
Note also, that this not only affect the Windows build, which is subject of this PR.

Holzhaus replied:
The idea was that people building from Mixxx probably want to do development and others that just grab the tarballs don't. But I'll just change it to use RelWithDebInfo in any case.

Holzhaus committed:
CMake: Default to RelWithDebInfo

daschuer reviewed (approved):
OK, that let's merge this. Thank you very much.
